### PR TITLE
docs/design: Support Spilling Unparalleled HashAgg

### PR DESCRIPTION
**Pull request by wshwsh12**
### What problem does this PR solve?

Issue Number: close #xxx 

Problem Summary:

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed: Design docs.

How it Works:

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:

### Release note 

- No release note

### CODE REVIEW

File: docs/design/2021-06-23-spilled-unparallel-hashagg.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
+<!--
+This is a template for TiDB's change proposal process, documented [here](./README.md).
+-->
+
+# Proposal: Support Spilling Unparalleled HashAgg
+
+- Author(s): [@wshwsh12](https://github.com/wshwsh12)
+- Discussion PR: Need Update
+- Tracking Issue: Need Update
```
Comment (mmyj) [inline docs/design/2021-06-23-spilled-unparallel-hashagg.md:9]: Tracking Issue: `https://github.com/pingcap/tidb/issues/25882`

File: docs/design/2021-06-23-spilled-unparallel-hashagg.md [hunk 2]
```diff
@@ -0,0 +1,70 @@
+<!--
+This is a template for TiDB's change proposal process, documented [here](./README.md).
+-->
+
+# Proposal: Support Spilling Unparalleled HashAgg
+
+- Author(s): [@wshwsh12](https://github.com/wshwsh12)
+- Discussion PR: Need Update
```
Comment (mmyj) [inline docs/design/2021-06-23-spilled-unparallel-hashagg.md:8]: Discussion PR: N/A

File: docs/design/2021-06-23-spilled-unparallel-hashagg.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
+<!--
+This is a template for TiDB's change proposal process, documented [here](./README.md).
```
Comment (mmyj) [inline docs/design/2021-06-23-spilled-unparallel-hashagg.md:2]: Should we remove this comment?

File: docs/design/2021-06-23-spilled-unparallel-hashagg.md [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+## Test Design
+
+### Functional Tests
+
+* Querying using aggregate functions should give correct result.
+
+### Scenario Tests
+
+* In a scenario where aggreagte executor use too many memory, this feature helps recude memory usage and run the sql successfully.
```
Comment (mmyj) [inline docs/design/2021-06-23-spilled-unparallel-hashagg.md:52]: Are `recude` typo?

File: docs/design/2021-06-23-spilled-unparallel-hashagg.md [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+* N/A
+
+### Benchmark Tests
+
+* The feature shouldn't cause any obvious performance regression (< 2%) on non-spilling scenario.
+
+## Impacts & Risks
+
+* Memory will still grow without increasing the number of new tuples in HashMap for distinct aggregate function.
```
Comment (mmyj) [inline docs/design/2021-06-23-spilled-unparallel-hashagg.md:64]: I don't understand this. It's seem that there is a contradiction between `memory still grow` and `without increasing the number of new tuples`.

Comment (wshwsh12) [reply docs/design/2021-06-23-spilled-unparallel-hashagg.md:60]: In fact, it is not contradict. For the distinct agg function, it is necessary to record those values that have appeared. We are using a set to record this information. This set will still grow during the aggregation process without increasing the number of new tuples in aggPartialResultMapper.
For example, 
```
type partialResult4CountDistinctInt struct {
	valSet set.Int64SetWithMemoryUsage
}
```

**Review by mmyj — APPROVED** [review]
lgtm

File: docs/design/2021-06-23-spilled-unparallel-hashagg.md [hunk 6]
```diff
@@ -0,0 +1,66 @@
+# Proposal: Support Spilling Unparalleled HashAgg
+
+- Author(s): [@wshwsh12](https://github.com/wshwsh12)
+- Discussion PR: N/A
```
Comment (tisonkun) [inline docs/design/2021-06-23-spilled-unparallel-hashagg.md:4]:
```suggestion
- Discussion PR: https://github.com/pingcap/tidb/pull/25792
```

File: docs/design/2021-06-23-spilled-unparallel-hashagg.md [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+* [Unresolved Questions](#unresolved-questions)
+
+## Introduction
+
+This proposal describes the implementation of Unparalleled HashAgg that can spill intermediate data when memory usage is higher than memory quota.
+Spilling for paralleled HashAgg will be supported later.
+
+## Motivation or Background
+
+Currently, the calculation logic of the aggregate executors in the TIDB is divided into two types, parallel and unparallel. However, when SQL memory usage exceeds the memory limit, neither implementation can use external memory to control memory usage and can only kill the SQL that is executing. In order to enable SQL to execute normally in the case of insufficient memory, we introduce the spilling algorithm for the unparallel aggreagte executor.
```
Comment (XuHuaiyu) [inline docs/design/2021-06-23-spilled-unparallel-hashagg.md:28]:
```suggestion
Currently, the calculation logic of the aggregate executors in the TiDB is divided into two types, parallel and unparallel. However, when SQL memory usage exceeds the memory quota, neither implementation can use external memory to control memory usage and can only kill the SQL that is executing. In order to enable SQL to execute normally in the case of insufficient memory, we introduce the spilling algorithm for the unparallel aggregate executor.
```

File: docs/design/2021-06-23-spilled-unparallel-hashagg.md [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+## Detailed Design
+
+In aggregate processing, memory increases when tuples are inserted into the hash table. So we can use the following algorithm to control the memroy incresaing:
```
Comment (XuHuaiyu) [inline docs/design/2021-06-23-spilled-unparallel-hashagg.md:32]:
```suggestion
In aggregate processing, memory increases when tuples are inserted into the hash table. So we can use the following algorithm to control the memory increasing:
```

File: docs/design/2021-06-23-spilled-unparallel-hashagg.md [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+## Test Design
+
+### Functional Tests
+
+* Querying using aggregate functions should give correct result.
+
+### Scenario Tests
+
+* In a scenario where aggregate executor use too many memory, this feature helps reduce memory usage and run the sql successfully.
```
Comment (XuHuaiyu) [inline docs/design/2021-06-23-spilled-unparallel-hashagg.md:48]: Some test cases also need to be added:
1. We also need a test for the parallel algorithm. When the parallel-agg exceeds the memory quota, the SQL will be canceled before. After the `agg-concurrency` args are set to 1, the SQL can run successfully.
2. SQLs like `select count(distinct a), sum(b) ...`

File: docs/design/2021-06-23-spilled-unparallel-hashagg.md [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+* The feature shouldn't cause any obvious performance regression (< 2%) on non-spilling scenario.
+
+## Impacts & Risks
+
+* Memory will still grow without increasing the number of new tuples in HashMap for distinct aggregate function.
+
+## Investigation & Alternatives
+
+## Future Work
+1. Support friendly spilling implement for distinct aggregate function.
```
Comment (XuHuaiyu) [inline docs/design/2021-06-23-spilled-unparallel-hashagg.md:65]:
```suggestion
1. Support friendly spilling implementation for the distinct aggregate functions.
```

File: docs/design/2021-06-23-spilled-unparallel-hashagg.md [hunk 11]
```diff
@@ -45,7 +45,10 @@ In aggregate processing, memory increases when tuples are inserted into the hash
 
 ### Scenario Tests
 
-* In a scenario where aggregate executor use too many memory, this feature helps reduce memory usage and run the sql successfully.
+* When the unparallel-agg exceeds the memory quota, this feature helps reduce memory usage and run the sql successfully.
+* When the parallel-agg exceeds the memory quota, the SQL will be canceled before. After the agg-concurrency args are set to 1, the SQL can run successfully.
+* When the ndv of the data is low, the SQL contains distinct function will be canceled before. After the agg-concurrency args are set to 1, the SQL can run successfully.
+* When the ndv of the data is high, the SQL contains distinct function will be canceled before. After the agg-concurrency args are set to 1, the SQL can be canceled successfully if there is insufficient memory.
```
Comment (XuHuaiyu) [inline docs/design/2021-06-23-spilled-unparallel-hashagg.md:51]: Why do we need to set the concurrency-related args when there exists an aggregation function with the keyword `distinct`.

Comment (wshwsh12) [reply docs/design/2021-06-23-spilled-unparallel-hashagg.md]: Addressed

**Review by XuHuaiyu — APPROVED** [review]
LGTM

**Pull request merged**